### PR TITLE
Fix incompatible behavior with TiDB when casting float into string

### LINKED ISSUE #9697: Incompatible behavior with TiDB when casting float into string
## Bug Report

Please answer these questions before submitting your issue. Thanks!

ref https://github.com/pingcap/tidb/issues/57652

### 4. What is your TiFlash version? (Required)

all

### PR DESCRIPTION
**Pull request by solotzg**
### What problem does this PR solve?

Issue Number: close #9697
Problem Summary:

- same problem in https://github.com/tikv/tikv/issues/16974, https://github.com/pingcap/tidb/issues/51109
- tidb tools (such as lightning) rely on crc function of tidb. crc module needs to cast basic type into string at first, which means all behavior about casting into str cannot be modified in order to keep backward compatibility.

### What is changed and how it works?

```commit-message
Fix incompatible behavior with TiDB when casting floating into string
- always convert the format into non-exponential representation
```

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/IO/WriteHelpers.cpp [hunk 1]
```diff
@@ -83,4 +85,167 @@ void writePointerHex(const void * ptr, WriteBuffer & buf)
...
+
+    Int64 exp_pos = sv.find(c_exp);
+    if (exp_pos < 0)
+    {
+        buf.write(sv.data(), sv.size());
+        return;
+    }
+
+    Int64 exp10 = 0;
+    if (exp_pos >= 0)
```
[the remark this answers is not in the document]
Comment (solotzg) [reply dbms/src/IO/WriteHelpers.cpp]: Fixed

**Review by windtalker — APPROVED** [review]
lgtm

**Review by yibin87 — APPROVED** [review]
LGTM

**Pull request merged**